packages/lodestar/src/network/reqResp.ts

```diff
@@ -175,70 +152,26 @@ export class ReqResp extends (EventEmitter as IReqEventEmitterClass) implements
     });
   };
 
-  private encodeRequest(method: Method, body: RequestBody): Buffer {
-    let output: Uint8Array;
-    switch (method) {
-      case Method.Status:
-        output = this.config.types.Status.serialize(body as Status);
-        break;
-      case Method.Goodbye:
-        output = this.config.types.Goodbye.serialize(body as Goodbye);
-        break;
-      case Method.BeaconBlocksByRange:
-        output = this.config.types.BeaconBlocksByRangeRequest.serialize(body as BeaconBlocksByRangeRequest);
-        break;
-      case Method.BeaconBlocksByRoot:
-        output = this.config.types.BeaconBlocksByRootRequest.serialize(body as BeaconBlocksByRootRequest);
-        break;
-    }
-    return Buffer.concat([
-      Buffer.from(varint.encode(output.length)),
-      output,
-    ]);
-  }
-
-  private decodeRequest(method: Method, data: Buffer): RequestBody {
-    const length = varint.decode(data);
-    const bytes = varint.decode.bytes;
-    if (
-      length !== data.length - bytes ||
-      length > MAX_CHUNK_SIZE
-    ) {
-      throw new Error(ERR_INVALID_REQ);
-    }
-    data = data.slice(bytes);
-    switch (method) {
-      case Method.Status:
-        return this.config.types.Status.deserialize(data);
-      case Method.Goodbye:
-        return this.config.types.Goodbye.deserialize(data);
-      case Method.BeaconBlocksByRange:
-        return this.config.types.BeaconBlocksByRangeRequest.deserialize(data);
-      case Method.BeaconBlocksByRoot:
-        return this.config.types.BeaconBlocksByRootRequest.deserialize(data);
-    }
-  }
   private async sendRequest<T extends ResponseBody>(
     peerInfo: PeerInfo,
     method: Method,
     body: RequestBody,
     requestOnly?: boolean
   ): Promise<T> {
-    const protocol = createRpcProtocol(method, this.encoding);
+    const protocol = createRpcProtocol(method, this.encoder.encoding);
     const {stream} = await this.libp2p.dialProtocol(peerInfo, protocol) as {stream: Stream};
     return await new Promise((resolve, reject) => {
       this.logger.verbose(`send ${method} request to ${peerInfo.id.toB58String()}`);
       const responseTimer = setTimeout(() => reject(new Error(ERR_RESP_TIMEOUT)), RESP_TIMEOUT);
       pipe(
-        [this.encodeRequest(method, body)],
+        [this.encoder.encodeRequest(method, body)],
         stream,
-        async (source: Promise<Buffer | {slice: () => Buffer}>[]) => {
+        this.encoder.decodeResponse(method),
+        async (source: AsyncIterable<ResponseBody>) => {
           try {
-            const responses = [];
-            for await (const val of source) {
-              const data = Buffer.isBuffer(val) ? val : val.slice();
-              const response = decodeResponseChunk(this.config, method, data);
-              responses.push(response.output);
+            const  responses = [];
+            for await (const response of source) {
```

Comment:
there should only ever be one response, no?

Reply:
In case of BlocksByRange, source will collect all decoded blocks

Reply:
Ok, so we're still exposing a ResponseBody (unchunked response) :(

Reply:
naming is wrong, ResponseChunk={err, output:ResponseBody}